### PR TITLE
feat(sdk): support redis aws

### PR DESCRIPTION
**Pull request by hasanaburayyan**
Not quite done yet but thought I would put it up incase anyone wanted an early peek.

- [x] tests
- [x] tighten up permissions
- [x] general tidy things

Closes: https://github.com/winglang/wing/issues/612

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingsdk/src/target-tf-aws/app.ts [hunk 1]
```diff
@@ -68,8 +83,126 @@ export class App extends CdktfApp {
...
+        Name: `${identifier}-vpc`,
+      },
+    });
+
+    // Create the subnets for the VPC, in order to ensure internet egress there
+    // is a minimum requirement of 2 subnets, one public and one private. As well
+    // as a NAT gateway and internet gateway. The NAT gateway is required to
+    // allow the private subnet to route traffic to the internet. The internet
+    // gateway is required to allow the NAT gateway to route traffic to the
+    // internet.
```
Comment (yoav-steinberg) [inline libs/wingsdk/src/target-tf-aws/app.ts:124]: Specifically for redis, why do we need internet access/EIP/public subnet etc..?

Comment (hasanaburayyan) [reply libs/wingsdk/src/target-tf-aws/app.ts:124]: For Redis we don’t  but for functions that will use Redis and be deployed into the private subnet we require this.

File: libs/wingsdk/src/target-tf-aws/redis.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import { App } from "./app";
+import { Function } from "./function";
+import { Code } from "../core";
+import * as core from "../core";
+import * as redis from "../redis";
+import { NameOptions, ResourceNames } from "../utils/resource-names";
+
+const ELASTICACHE_NAME_OPTS: NameOptions = {
+  maxLen: 50,
+  disallowedRegex: /([^a-z0-9]+)/g,
```
Comment (Chriscbr) [inline libs/wingsdk/src/target-tf-aws/redis.ts:15]: I'm guessing we can allow upper case characters here, as long as we apply the right case conversion:
```suggestion
  disallowedRegex: /([^a-zA-Z0-9]+)/g,
  case: CaseConventions.LOWERCASE,
```

(this API still needs some work I think)

File: libs/wingsdk/src/target-tf-aws/redis.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    this.subnet = app.subnets.private;
+    const clusterName = ResourceNames.generateName(this, ELASTICACHE_NAME_OPTS);
+
+    this.securityGroup = new SecurityGroup(this, "securityGroup", {
+      vpcId: vpc.id,
+      name: `${this.node.addr.slice(-8)}-securityGroup`,
+      ingress: [
+        {
+          cidrBlocks: [this.subnet.cidrBlock],
+          fromPort: 6379,
```
Comment (Chriscbr) [inline libs/wingsdk/src/target-tf-aws/redis.ts:53]: Make this a constant
```suggestion
          fromPort: REDIS_PORT,
```

File: libs/wingsdk/test/target-tf-aws/redis.inflight.test.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+});
+
+const fakeRedisClient = {
+  items: new Map<any, any>(),
+  set: (key: string, value: string) => {
+    fakeRedisClient.items.set(key, value);
+  },
+  get: (key: string) => {
+    return fakeRedisClient.items.get(key);
+  },
```
Comment (Chriscbr) [inline libs/wingsdk/test/target-tf-aws/redis.inflight.test.ts:99]: Wow, it's like we have a whole redis simulation right here 😀

**Review by Chriscbr — APPROVED** [review]
Fantastic work! 🚀

**Pull request merged**